src/core/server/ui_settings/ui_settings_client.ts

```diff
@@ -201,16 +228,45 @@ export class UiSettingsClient implements IUiSettingsClient {
     return filteredValues;
   }
 
+  /**
+   * group change into different scopes
+   * @param changes ui setting changes
+   * @returns [global, user]
+   */
+  private groupChanges(changes: Record<string, any>) {
+    const userLevelKeys = [] as string[];
+    Object.entries(this.defaults).forEach(([key, value]) => {
+      if (value.scope === 'user' || (Array.isArray(value.scope) && value.scope.includes('user'))) {
+        userLevelKeys.push(key);
+      }
+    });
+    const userChanges = {} as Record<string, any>;
+    const globalChanges = {} as Record<string, any>;
+
+    Object.entries(changes).forEach(([key, val]) => {
+      if (userLevelKeys.includes(key)) {
```

Comment:
Here is just a few changes to demonstrate what I was referring:
```diff
diff --git a/src/core/server/ui_settings/ui_settings_client.ts b/src/core/server/ui_settings/ui_settings_client.ts
index e78f797465..b48a32e6d9 100644
--- a/src/core/server/ui_settings/ui_settings_client.ts
+++ b/src/core/server/ui_settings/ui_settings_client.ts
@@ -50,6 +50,7 @@ export interface UiSettingsServiceOptions {
 interface ReadOptions {
   ignore401Errors?: boolean;
   autoCreateOrUpgradeIfMissing?: boolean;
+  scope?: Scope;
 }
 
 interface UserProvidedValue<T = unknown> {
@@ -62,6 +63,8 @@ type UiSettingsRawValue = UiSettingsParams & UserProvidedValue;
 type UserProvided<T = unknown> = Record<string, UserProvidedValue<T>>;
 type UiSettingsRaw = Record<string, UiSettingsRawValue>;
 
+type Scope = 'user' | 'global';
+
 export class UiSettingsClient implements IUiSettingsClient {
   private readonly type: UiSettingsServiceOptions['type'];
   private readonly id: UiSettingsServiceOptions['id'];
@@ -100,13 +103,13 @@ export class UiSettingsClient implements IUiSettingsClient {
     return this.defaults[key]?.value;
   }
 
-  async get<T = any>(key: string): Promise<T> {
-    const all = await this.getAll();
+  async get<T = any>(key: string, scope?: Scope): Promise<T> {
+    const all = await this.getAll(scope);
     return all[key];
   }
 
-  async getAll<T = any>() {
-    const raw = await this.getRaw();
+  async getAll<T = any>(scope?: Scope) {
+    const raw = await this.getRaw(scope);
 
     return Object.keys(raw).reduce((all, key) => {
       const item = raw[key];
@@ -115,8 +118,10 @@ export class UiSettingsClient implements IUiSettingsClient {
     }, {} as Record<string, T>);
   }
 
-  async getUserProvided<T = unknown>(): Promise<UserProvided<T>> {
-    const userProvided: UserProvided<T> = this.onReadHook<T>(await this.read());
+  async getUserProvided<T = unknown>(scope?: Scope): Promise<UserProvided<T>> {
+    const userProvided: UserProvided<T> = this.onReadHook<T>(
+      await this.read({ scope, autoCreateOrUpgradeIfMissing: scope === 'global' })
+    );
 
     // write all overridden keys, dropping the userValue is override is null and
     // adding keys for overrides that are not in saved object
@@ -128,13 +133,13 @@ export class UiSettingsClient implements IUiSettingsClient {
     return userProvided;
   }
 
-  async setMany(changes: Record<string, any>) {
+  async setMany(changes: Record<string, any>, scope?: Scope) {
     this.onWriteHook(changes);
-    await this.write({ changes });
+    await this.write({ changes, scope });
   }
 
-  async set(key: string, value: any) {
-    await this.setMany({ [key]: value });
+  async set(key: string, value: any, scope?: Scope) {
+    await this.setMany({ [key]: value }, scope);
   }
 
   async remove(key: string) {
@@ -159,8 +164,8 @@ export class UiSettingsClient implements IUiSettingsClient {
     }
   }
 
-  private async getRaw(): Promise<UiSettingsRaw> {
-    const userProvided = await this.getUserProvided();
+  private async getRaw(scope?: Scope): Promise<UiSettingsRaw> {
+    const userProvided = await this.getUserProvided(scope);
     return defaultsDeep(userProvided, this.defaults);
   }
 
@@ -204,25 +209,32 @@ export class UiSettingsClient implements IUiSettingsClient {
   private async write({
     changes,
     autoCreateOrUpgradeIfMissing = true,
+    scope = 'global',
   }: {
     changes: Record<string, any>;
     autoCreateOrUpgradeIfMissing?: boolean;
+    scope?: Scope;
   }) {
     changes = this.translateChanges(changes, 'timeline', 'timelion');
+    const id = this.getId(scope);
     try {
-      await this.savedObjectsClient.update(this.type, this.id, changes);
+      await this.savedObjectsClient.update(this.type, id, changes);
     } catch (error) {
       if (!SavedObjectsErrorHelpers.isNotFoundError(error) || !autoCreateOrUpgradeIfMissing) {
         throw error;
       }
 
-      await createOrUpgradeSavedConfig({
-        savedObjectsClient: this.savedObjectsClient,
-        version: this.id,
-        buildNum: this.buildNum,
-        log: this.log,
-        handleWriteErrors: false,
-      });
+      if (scope === 'global') {
+        await createOrUpgradeSavedConfig({
+          savedObjectsClient: this.savedObjectsClient,
+          version: this.id,
+          buildNum: this.buildNum,
+          log: this.log,
+          handleWriteErrors: false,
+        });
+      } else {
+        await this.savedObjectsClient.create('config', {}, { id });
+      }
 
       await this.write({
         changes,
@@ -231,12 +243,21 @@ export class UiSettingsClient implements IUiSettingsClient {
     }
   }
 
+  getId(scope?: Scope) {
+    if (scope === 'user') {
+      return '__USER_SETTING_FLAG__';
+    }
+    return this.id;
+  }
+
   private async read({
     ignore401Errors = false,
     autoCreateOrUpgradeIfMissing = true,
+    scope = 'global',
   }: ReadOptions = {}): Promise<Record<string, any>> {
+    const id = this.getId(scope);
     try {
-      const resp = await this.savedObjectsClient.get<Record<string, any>>(this.type, this.id);
+      const resp = await this.savedObjectsClient.get<Record<string, any>>(this.type, id);
       return this.translateChanges(resp.attributes, 'timelion', 'timeline');
     } catch (error) {
       if (SavedObjectsErrorHelpers.isNotFoundError(error) && autoCreateOrUpgradeIfMissing) {
@@ -262,6 +283,10 @@ export class UiSettingsClient implements IUiSettingsClient {
         return {};
       }
 
+      if (scope === 'user') {
+        return {};
+      }
+
       throw error;
     }
   }
```

Reply:
> What if a setting is both `global` and `user`? This implementation implies a setting can only be `global` or `user`.
> 
> This is what I proposed in the RFC, any reason why this didn't work?
> 
> ## Reading a setting with scope
> When calling uiSettings.get(key), it returns the value based on the following rules:
> 
> 1. If the setting was registered with scope global or empty, the global setting will be returned.
> 2. If the setting was registered with scope user, the user setting will be returned
> 3. If the setting was registered with both user and global, the user setting will be returned if it's set, otherwise, return the global setting
> 4. If calling uiSettings.get(key, defaultValue, 'user') with scope specified, it will return the setting of the specified scope. If the value doesn't not exist, then return the defaultValue. This requires to update the current uiSettings service's interface
> 
> ## Writing a setting with scope
> When calling uiSettings.set(key, value), it follows the following rules:
> 
> 1. If the setting was registered with scope global or empty, it will write the value to global setting saved object.
> 2. If the setting was registered with scope user, it will write the value to user setting saved object.
> 3. If the setting was registered with both user and global, it will by default update the user setting saved object.
> 4. If calling uiSettings.set(key, value, 'global') with scope specified, it will write the value to the corresponding saved object. This requires to update the current uiSettings service's interface

If a setting is both `global` and `user`, it will set to user setting by default, when get it they will merge, and returns same value when set. 

This PR implementation didn't implement item 4 for both read and write with specified scope that requires client to know which setting is global and which is user. I agree that will give more flexibility to control how we are going to store the setting but that also need additional validation make sure changes are align with specified scope.


To support item 4, that will be similar change as you suggested plus router accept scope parameter and scope validation.

As the default behavior could handle most of case, that's the reason i didn't implement it this PR.

Reply:
scope added to both get and set method